File: library/src/main/java/net/robinx/queue/Job.java
package net.robinx.queue;

/**
 * Created by Robin on 2016/6/28 16:52.
 */
public abstract class Job<T extends Job<T>> implements java.lang.Comparable<T> {
    protected Object tag;

    protected boolean isCancel;
    protected boolean isFinish;

    protected Integer mSequence;
    protected Priority mPriority = Priority.NORMAL;

    public Object getTag() {
        return tag;
    }

    public Job tag(Object tag) {
        this.tag = tag;
        return this;
    }

    public Priority getPriority() {
        return mPriority;
    }

    public Job priority(Priority priority) {
        mPriority = priority;
        return this;
    }

    public boolean isCancel() {
        return isCancel;
    }

    public Job cancel() {
        isCancel = true;
        return this;
    }

    public boolean isFinish() {
        return isFinish;
    }

    public Job finish() {
        isFinish = true;
        return this;
    }

    public Integer getSequence() {
        return mSequence;
    }

    public Job sequence(Integer sequence) {
        mSequence = sequence;
        return this;
    }

    @Override
    public int compareTo(T another) {
        Priority left = this.getPriority();
        Priority right = another.getPriority();
        return left == right ? this.mSequence - another.mSequence : right.ordinal() - left.ordinal();
    }

    public abstract Object getRepeatFilterKey();
}

File: library/src/main/java/com/github/jobqueue/Job.kt
package com.github.jobqueue

/**
 * Created by Robin on 2016/6/28 16:52.
 */
abstract class Job<T : Job<T>> : Comparable<T> {
    var tag: Any? = null
        protected set

    var isCancel: Boolean = false
        protected set
    var isFinish: Boolean = false
        protected set

    var sequence: Int? = null
        protected set
    var priority: Priority = Priority.NORMAL
        protected set

    fun tag(tag: Any?): Job<*> {
        this.tag = tag
        return this
    }

    fun priority(priority: Priority): Job<*> {
        this.priority = priority
        return this
    }

    fun cancel(): Job<*> {
        isCancel = true
        return this
    }

    fun finish(): Job<*> {
        isFinish = true
        return this
    }

    fun sequence(sequence: Int?): Job<*> {
        this.sequence = sequence
        return this
    }

    override fun compareTo(another: T): Int {
        val left = this.priority
        val right = another!!.priority
        return if (left == right) sequence!! - another.sequence!! else right.ordinal - left.ordinal
    }

    abstract val repeatFilterKey: Any?
}
